File: practice/src/main/java/com/decade/practice/web/advices/ExceptionControllerAdvice.java
package com.decade.practice.web.advices;

import jakarta.persistence.EntityNotFoundException;
import jakarta.persistence.OptimisticLockException;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.handler.annotation.MessageExceptionHandler;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.method.annotation.HandlerMethodValidationException;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.NoSuchElementException;

@RestControllerAdvice
public class ExceptionControllerAdvice extends ResponseEntityExceptionHandler {

      @ExceptionHandler({
            EntityNotFoundException.class,
            NoSuchElementException.class,
            NullPointerException.class
      })
      @ResponseStatus(value = HttpStatus.NOT_FOUND)
      @MessageExceptionHandler({
            EntityNotFoundException.class,
            NoSuchElementException.class,
            NullPointerException.class
      })
      public void handleNoElement(Exception e) {
            e.printStackTrace();
      }

      @ExceptionHandler(OptimisticLockException.class)
      @ResponseStatus(value = HttpStatus.CONFLICT)
      public void handleLockException() {
      }

      @Override
      protected ResponseEntity<Object> handleHandlerMethodValidationException(
            HandlerMethodValidationException ex,
            HttpHeaders headers,
            HttpStatusCode status,
            WebRequest request
      ) {
            return ResponseEntity
                  .badRequest()
                  .body(
                        ex.getAllErrors()
                              .get(0).getDefaultMessage()
                  );
      }

      @Override
      protected ResponseEntity<Object> handleMethodArgumentNotValid(
            MethodArgumentNotValidException ex,
            HttpHeaders headers,
            HttpStatusCode status,
            WebRequest request
      ) {
            return ResponseEntity
                  .badRequest()
                  .body(ex.getMessage());
      }
}

File: practice/src/main/java/com/decade/practice/controllers/advices/ExceptionControllerAdvice.kt
package com.decade.practice.controllers.advices

import jakarta.persistence.EntityNotFoundException
import jakarta.persistence.OptimisticLockException
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.messaging.handler.annotation.MessageExceptionHandler
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import org.springframework.web.method.annotation.HandlerMethodValidationException
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler


@RestControllerAdvice
class ExceptionControllerAdvice : ResponseEntityExceptionHandler() {

      @ExceptionHandler(
            EntityNotFoundException::class, NoSuchElementException::class, NullPointerException::class
      )
      @ResponseStatus(value = HttpStatus.NOT_FOUND, reason = "THE REQUESTED RESOURCE NOT FOUND")
      @MessageExceptionHandler(
            EntityNotFoundException::class, NoSuchElementException::class, NullPointerException::class
      )
      fun handleNoElement(e: Exception) {
            e.printStackTrace()
      }

      @ExceptionHandler(OptimisticLockException::class)
      @ResponseStatus(value = HttpStatus.BAD_REQUEST, reason = "UPDATE FAILED, TRY AGAIN")
      fun handleLockException() {
      }

      override fun handleHandlerMethodValidationException(
            ex: HandlerMethodValidationException,
            headers: HttpHeaders,
            status: HttpStatusCode,
            request: WebRequest
      ): ResponseEntity<Any> {
            return ResponseEntity.badRequest().body(
                  ex.allErrors
                        .first().defaultMessage
            )
      }

      override fun handleMethodArgumentNotValid(
            ex: MethodArgumentNotValidException,
            headers: HttpHeaders,
            status: HttpStatusCode,
            request: WebRequest
      ): ResponseEntity<Any> {
            return ResponseEntity.badRequest().body(ex.message)
      }
}